common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -350,6 +354,14 @@ static WeakOrderQueue allocate(Stack<?> stack, Thread thread) {
         void add(DefaultHandle<?> handle) {
             handle.lastRecycledId = id;
 
+            // While we also enforce the recycling ratio one we transfer objects from the WeakOrderQueue to the Stack
+            // we better should enforce it as well early. Missing to do so may let the WeakOrderQueue grow very fast
+            // without control if the Stack
+            if ((++handleRecycleCount & ratioMask) != 0) {
```

Comment:
Since determine if we need to discard the recycle object(`(++handleRecycleCount & ratioMask) != 0`) needs to be executed in both `pushNow` and `pushLater`, is it a better idea to advance it to the `Stack#push`?

Reply:
The reason why I didn't do this was that I still like to have the validation done first in `pushNow(...)`

Reply:
Emmm...
A little confused. Seems that `pushNow` and `pushLater` is in two different branch. 

Not like this:
```
if (pushNow) {
    pushLater
}
```

Reply:
@carryxyh not sure I get what you suggest... Can you maybe just open a PR or show a diff here ?

Reply:
Ok, have a try later.

Reply:
@normanmaurer 
A temp branch here: https://github.com/netty/netty/pull/9732

Reply:
As commented in your PR what you propose is not thread-safe and will not work. So lets keep this pr as it is imho... WDYT ?

Reply:
I ignored the thread safety. Thanks for reminding.  :)